controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,887 @@
+  int FreeCt() { return N - 1 - FullCt(); }
+
+  // Get the oldest byte from the buffer.
+  // Returns -1 if the buffer is empty
+  int Get() {
+    int ret = -1;
+
+    bool p = IntSuspend();
+
+    if (head != tail) {
+      ret = buff[tail++];
+      if (tail >= N)
+        tail = 0;
+    }
+
+    IntRestore(p);
+    return ret;
+  }
+
+  // Add a byte to the buffer
+  // Returnes true on success, false if the buffer is full
+  bool Put(uint8_t dat) {
+    bool ret = false;
+
+    bool p = IntSuspend();
+
+    int h = head + 1;
+    if (h >= N)
+      h = 0;
+
+    if (h != tail) {
+      buff[head] = dat;
+      head = h;
+      ret = true;
+    }
+
+    IntRestore(p);
+
+    return ret;
+  }
+};
+
+class UART {
+  CircBuff<128> rxDat;
+  CircBuff<128> txDat;
+  UART_Regs *const reg;
+
+public:
+  UART(UART_Regs *const r) : reg(r) {}
+
+  void Init(int baud) {
+
+    // Set baud rate register
+    reg->baud = CPU_FREQ / baud;
+
+    // Enable the UART and receive interrupts
+    reg->ctrl[0] = 0x002D;
+  }
+
+  void ISR() {
+
+    // Check for over run error and framing errors.
+    // Clear those errors if they're set to avoid
+    // further interrupts from them.
+    if (reg->status & 0x000A)
+      reg->intClear = 0x000A;
+
+    // See if we received a new byte
+    if (reg->status & 0x0020)
+      rxDat.Put(reg->rxDat);
+
+    // Check for transmit data register empty
+    if (reg->status & reg->ctrl[0] & 0x0080) {
+      int ch = txDat.Get();
+
+      // If there's nothing left in the transmit buffer,
+      // just disable further transmit interrupts.
+      if (ch < 0)
+        reg->ctrl[0] &= ~0x0080;
+
+      // Otherwise, Send the next byte
+      else
+        reg->txDat = ch;
+    }
+  }
+
+  // Read up to len bytes and store them in the passed buffer.
+  // Returns the number of bytes actually read.
+  uint16_t read(char *buf, uint16_t len) {
+
+    for (uint16_t i = 0; i < len; i++) {
+      int ch = rxDat.Get();
+      if (ch < 0)
+        return i;
+      *buf++ = ch;
+    }
+    return len;
+  }
+
+  // Write up to len bytes to the buffer and return the
+  // number actually written.
+  uint16_t write(const char *buf, uint16_t len) {
+
+    uint16_t i;
+    for (i = 0; i < len; i++) {
+      if (!txDat.Put(*buf++))
+        break;
+    }
+
+    // Enable the tx interrupt
+    reg->ctrl[0] |= 0x0080;
+
+    return i;
+  }
+
+  // Return the number of bytes currently in the
+  // receive buffer and ready to be read.
+  int RxFull() { return rxDat.FullCt(); }
+
+  // Returns the number of free locations in the
+  // transmit buffer.
+  int TxFree() { return txDat.FreeCt(); }
+};
+
+static UART rpUART(UART3_BASE);
+
+// The UART that talks to the rPi uses the following pins:
+//    PB10 - TX
+//    PB11 - RX
+//    PB13 - RTS
+//    PB14 - CTS
+//
+// These pins are connected to UART3
+static void InitUARTs() {
+  // NOTE - The UART functionality hasn't been tested due to lack of hardware!
```

Comment:
I'm expecting the PCB Monday or Tuesday.  The first thing I'm going to be doing is testing out the hardware (PCB) to make sure everything works as expected.  Part of that will be testing the UART.
This UART is the link between the RPI and the controller.  It's inconceivable that it could not work and no one would notice.
I can add an issue (or you could) if you really think it's necessary, but seems silly.